### PR TITLE
[Wisp] Blocking io support for SocketChannel and Pipe

### PR DESCRIPTION
**Pull request by joeyleeeeeee97**
Summary: Hook blocking entry at pipeChannel and SocketChannel,
current WispTask would block itself to wait for next scheduling.

Test Plan: jtreg BlockIngIOTest.

Reviewed-by: zhengxiaolinX, yuleil

Issue: https://github.com/alibaba/dragonwell8/issues/143

### CODE REVIEW

File: src/share/classes/java/nio/channels/spi/AbstractSelectableChannel.java [hunk 1]
```diff
@@ -25,15 +25,12 @@
...
 import java.io.IOException;
-import java.nio.channels.CancelledKeyException;
-import java.nio.channels.ClosedChannelException;
-import java.nio.channels.ClosedSelectorException;
-import java.nio.channels.IllegalBlockingModeException;
-import java.nio.channels.IllegalSelectorException;
-import java.nio.channels.SelectableChannel;
-import java.nio.channels.SelectionKey;
-import java.nio.channels.Selector;
+import java.nio.channels.*;
```
Comment (yuleil) [inline src/share/classes/java/nio/channels/spi/AbstractSelectableChannel.java:33]: unnecessary change

File: src/share/classes/sun/nio/ch/ServerSocketAdaptor.java [hunk 2]
```diff
@@ -34,11 +36,7 @@
 import java.net.SocketException;
 import java.net.SocketTimeoutException;
 import java.net.StandardSocketOptions;
-import java.nio.channels.ClosedChannelException;
-import java.nio.channels.IllegalBlockingModeException;
-import java.nio.channels.NotYetBoundException;
-import java.nio.channels.ServerSocketChannel;
-import java.nio.channels.SocketChannel;
+import java.nio.channels.*;
```
Comment (yuleil) [inline src/share/classes/sun/nio/ch/ServerSocketAdaptor.java:39]: unnecessary change

File: src/linux/classes/com/alibaba/wisp/engine/WispEngine.java [hunk 3]
```diff
@@ -344,6 +344,25 @@ public void getCpuTime(long[] ids, long[] times) {
                     }
                 }
             }
+
+            @Override
+            public int poll(SelectableChannel channel, int interestOps, long millsTimeOut) throws IOException {
```
Comment (yuleil) [inline src/linux/classes/com/alibaba/wisp/engine/WispEngine.java:349]: 1. WAE.poll should return polled event count;
2. WAE.poll receive NET.* events represent as parameters will be clearer.
3. Coroutines are not interrupted when the socket closes. Leave a TODO there

Comment (joeyleeeeeee97) [reply src/linux/classes/com/alibaba/wisp/engine/WispEngine.java:356]: OK. please have another look.

Close socket is ok here, because we would check isOpen() every time.

File: src/linux/classes/com/alibaba/wisp/engine/WispEngine.java [hunk 4]
```diff
@@ -344,6 +346,40 @@ public void getCpuTime(long[] ids, long[] times) {
...
+                    return millsTimeOut > 0 && task.timeOut.expired() ? 0 : 1;
+                } finally {
+                    if (millsTimeOut > 0) {
+                        task.carrier.cancelTimer();
+                    }
+                    unregisterEvent();
+                }
+            }
+
+            int translateToSelectionKey(int event) {
```
Comment (joeyleeeeeee97) [inline src/linux/classes/com/alibaba/wisp/engine/WispEngine.java:375]: Can't use switch here because Net.POLLIN is not constant.

File: src/share/classes/sun/nio/ch/SocketAdaptor.java [hunk 5]
```diff
@@ -25,6 +25,8 @@
 
 package sun.nio.ch;
 
+import com.alibaba.wisp.engine.WispEngine;
```
Comment (yuleil) [inline src/share/classes/sun/nio/ch/SocketAdaptor.java:28]: unnecessary change

File: src/linux/classes/com/alibaba/wisp/engine/WispEngine.java [hunk 6]
```diff
@@ -344,6 +346,40 @@ public void getCpuTime(long[] ids, long[] times) {
...
                 }
             }
+
+            @Override
+            public int poll(SelectableChannel channel, int interestOps) throws IOException {
+                return poll(channel, interestOps, -1);
+            }
+
+            @Override
+            public int poll(SelectableChannel channel, int interestOps, long millsTimeOut) throws IOException {
```
Comment (yuleil) [inline src/linux/classes/com/alibaba/wisp/engine/WispEngine.java:356]: add JavaDoc to point out it's a Net.* event represent

File: src/share/classes/sun/nio/ch/SocketAdaptor.java (hunk 5 above)
Comment (joeyleeeeeee97) [reply src/share/classes/sun/nio/ch/SocketAdaptor.java]: ok

File: src/linux/classes/com/alibaba/wisp/engine/WispEngine.java [hunk 7]
```diff
@@ -344,6 +346,42 @@ public void getCpuTime(long[] ids, long[] times) {
...
             }
+
+            /**
+             * @param channel Blocking SocketChannel waiting for interestOps
+             * @param interestOps Net.* enum constant interests
+             * @param millsTimeOut timeout in milliseconds
+             * @return active event count
+             * @throws IOException
+             */
+            @Override
```
Comment (yuleil) [inline src/linux/classes/com/alibaba/wisp/engine/WispEngine.java:357]: put the java doc above method in the interface

File: src/share/classes/sun/nio/ch/ServerSocketAdaptor.java [hunk 8]
```diff
@@ -25,6 +25,7 @@
 
 package sun.nio.ch;
 
+
```
Comment (yuleil) [inline src/share/classes/sun/nio/ch/ServerSocketAdaptor.java:28]: unnecessary

File: src/share/classes/sun/nio/ch/ServerSocketChannelImpl.java [hunk 9]
```diff
@@ -278,7 +271,9 @@ public SocketChannel accept() throws IOException {
             if (n < 1)
                 return null;
 
-            IOUtil.configureBlocking(newfd, true);
+            if (!WispEngine.transparentWispSwitch()) {
```
Comment (yuleil) [inline src/share/classes/sun/nio/ch/ServerSocketChannelImpl.java:274]: use configureAsNonBlockingForWisp

File: src/share/classes/sun/nio/ch/SocketAdaptor.java [hunk 10]
```diff
@@ -25,6 +25,7 @@
 
 package sun.nio.ch;
 
+
```
Comment (yuleil) [inline src/share/classes/sun/nio/ch/SocketAdaptor.java:28]: unnecessary

File: src/solaris/classes/sun/nio/ch/SourceChannelImpl.java [hunk 11]
```diff
@@ -80,6 +85,11 @@ public int getFDVal() {
         this.fd = fd;
         this.fdVal = IOUtil.fdVal(fd);
         this.state = ST_INUSE;
+        try {
+            configureAsNonBlockingForWisp(fd);
+        } catch (IOException e) {
+            throw new RuntimeException("Unexpected error at configureAsNonBlockingForWisp", e);
```
Comment (yuleil) [inline src/solaris/classes/sun/nio/ch/SourceChannelImpl.java:91]: UncheckIOException

File: src/solaris/classes/sun/nio/ch/SourceChannelImpl.java [hunk 12]
```diff
@@ -165,6 +175,12 @@ public int read(ByteBuffer dst) throws IOException {
                 thread = NativeThread.current();
                 do {
                     n = IOUtil.read(fd, dst, -1, nd);
+                    if (isBlocking() && WispEngine.transparentWispSwitch()) {
```
Comment (yuleil) [inline src/solaris/classes/sun/nio/ch/SourceChannelImpl.java:178]: WispEngine.transparentWispSwitch() && isBlocking()

Comment (joeyleeeeeee97) [reply src/solaris/classes/sun/nio/ch/SourceChannelImpl.java:178]: ok

**Review by yuleil — APPROVED** [review]

**Review by zhengxiaolinX — APPROVED** [review]
This patch looks good to me.
